---
title: "TP bioinfo IFSBM UE Big Data"
subtitle: "Modèles prédictifs"
author: "Loic Verlingue, Md, PhDc loic.verlingues@gustaveroussy.fr"
output: html_notebook
---

---
# What you will do in this R Notebook

## Introduction
With this notebook you will work on real cancer genomic data to address clinical questions.  
The problem can be summarized this way: your input is a huge amount of variables (genes, DNA or RNA sequences per patients) and you want to provide the clinician a simple output that he can use for his patients.

The global concept is a reduction of information for human understanding: a central objective in many machine learning projects.

This is actually one of the current main challenges of personalized medecine, so I'll hope you'll enjoye this Notebook and that it can help you to adress your current and/or future research challenges.

## Main steps
I'll guide you trhough the aquisition of TCGA data. We will need to reduce the queries to most relevant genes otherwise your computer won't be able to handle the job. 

Then I'll will explain to you a simple modeling strategy for classification, that works very well. Finally, if we have time, I'll introduce to you survival analysis and provide you a glimpse on the difficulties and challenges of such analysis. 

It is where the current science stands! To go further you'll have to become a researcher in bioinformatics and machine learning. 

## In practice
The code is provided to you. You will just have to follow the instructions all along the Notebook.
I'll give you a little bit of explanationn when needed. 

Pay attention: there will be some questions to answer in this notebook.

Finally you will have a partial code to complete at the end of the notebook, and give it back to us.

*--> lets' go!*

---
# Get familiar with R Notebook

### Execute code

```{r}
 #Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Enter*. 
print("Welcome to 'Big Data et modèles prédictifs' practical session")
set.seed(1234)
```

### Edit your Notebook

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).


# Load your material

## a short list of important genes
The Cancer Gene Census, CGS, is a selection of important genes for cancer devellopment. 
To load the [Cancer Gene Census](https://cancer.sanger.ac.uk/census) table you need a login. To facilitate your work, I have provided it to you. You simply have to run the following cell.

```{r}
urlfile<-"../data/CancerGeneCensusCOSMIC.csv"
CGS<-read.csv(urlfile,stringsAsFactors = F)
GENES<-CGS$Gene.Symbol[CGS$Hallmark=="Yes"]
print(paste("You have selected",length(GENES),"important cancer genes like:", paste(sample(GENES,20), collapse = ", "), "..."))
```

## Data from TCGA

Now you need data from [TCGA](https://cancergenome.nih.gov/)!

The TCGA is a database containing clinical and molecular data (Whole Exome, RNASeq, microarray and others) and clinical information such as survival from several thousand American patients. This anonymized data is accessible: to access the raw data you will need the authorization of a data user commitee. Instead the processed data is free of access. It can nevertheless be difficult to retrieve this data by hand and then load it into R. 

For you, I have built a function that will load the data you need from a list of genes and cancer types. The processed data comes from [cBioportal](http://www.cbioportal.org/), a reference  web site widely used for the exploration and downloading of TCGA data.
For that you will need to install the official cBioportal R package [cgdsr](http://www.cbioportal.org/rmatlab) and my function [LoadcBioportal150119.R](https://github.com/DITEP/TPBigDataIFSBM) (if you are curious you can go and check the source code of this function).

```{r}
source("../../lib/LoadcBioportal150119.R")
```

You will also need some other usefull R packages such as:

- the [glmnet](https://cran.r-project.org/web/packages/glmnet/glmnet.pdf) package to compute linear models

- the [survival](https://cran.r-project.org/web/packages/survival/survival.pdf) package to analyse survival data

```{r}
# Load usefull packages
library(reshape2)
library(glmnet)
library(survival)
```

# Use LoadcBioportal() function

Your first step is to use the LoadcBioportal() function to load the TCGA data.
A R function is composed of several arguments that you may change depending on your task.
For example here you can query for various tumor types existing in the TCGA. Be careful, you can use either the official
TCGA nomeclature (luad = lung adenocarcinoma, gbm = glioblastoma), or the name of the organ of origin (lung = luad & lusc, brain = gbm & lgg).

You will also query only the important genes GENES selected before (otherwise it could be very long).

In regular expression (=computer langage), the "OR" logical connector is writen `|`.
Moreoever, `T` stands for `TRUE` and `F` stands for `FALSE`.
Look at the arguments to understand the possibilities of this function.

```{r}
#############
# load data
#############
TCGAdata<-LoadcBioportal(
  Genes=GENES,
  Organ="luad|gbm",
  ClinicNeeded=T,
  MutNeeded=F,
  RNANeeded=T,
  NormalizeRNA=T,
  FunctionalAnnot=F,
  PDF=F,
  Tests=T
)
STUDY<-TCGAdata$STUDY
print("Here are your patients' numbers:")
print(table(TCGAdata$CLINIC$study))
```


This function returns a list called `TCGAdata` where each request is stored under the following names: 

```{r}
names(TCGAdata)
```


*rescue only, else skip*

```{r}
# EXP<-read.csv2("https://github.com/gustaveroussy/IFSBM-bigdata/tree/master/TP_IFSBM_module12/2020/labs/lab_2/data/EXPlung_gbm.csv", row.names = 1)
# CLINIC<-read.csv2("https://github.com/gustaveroussy/IFSBM-bigdata/tree/master/TP_IFSBM_module12/2020/labs/lab_2/data/Cliniclung_gbm.csv", row.names = 1)
# TCGAdata<-list()
# TCGAdata["EXP"]<-EXP
# TCGAdata["CLINIC"]<-CLINIC
```

*end rescue only*

It is of good practice to check the dimension of your data. The dimensions are in the form [N rows, N columns].

```{r}
# check dimensions of your data                          
lapply(TCGAdata,dim)
```

See you didn't load mutation data because of the argument `MutNeeded = F` so the dimensions of the `TCGAdata$MUT` is [0,0].
The function has selected and sorted the patients so that you have a `TCGAdata$EXP` and `TCGAdata$CLINIC` with the same patients as rows and no NAs.

# Description of the data

The first very important step is to visualize and describe your data. It can be tricky when you have huge multidimensional matrices. 

## Distribution with densities
For continuous data such as RNAseq, simply plotting the distribution is a good first step. See that the LoadcBioportal()
function has performed some processing of the RNA data because you have set the argument `NormalizeRNA = T`

```{r}
# check distribution of gene expression
hist(as.matrix(TCGAdata$EXP[TCGAdata$CLINIC$study=="luad",]),col=2,
     main="Distribution of gene expressions",xlab="Normalized TPM", freq = F)
hist(as.matrix(TCGAdata$EXP[TCGAdata$CLINIC$study=="gbm",]),add=T,col=adjustcolor(3,0.7), freq = F)
legend("topright",legend=c("luad","gbm"), fill=2:3, cex=0.8)
```

### Quizz

1) What is TPM? 

your answer here: 

2) What is the distribution of your RNAseq data?

your answer here: 

## Distribution with Singular Value Decomposition or Principal Component Analysis

There is another way to know if the distribution of the gene expressions are very different between independant studies.

```{r}
# compute and plot svd in 2 dimensions
res.svd<-svd(TCGAdata$EXP, nu = 2, nv = 2)
plot(res.svd$v[,1],res.svd$v[,2],col=ifelse(TCGAdata$CLINIC$study=="luad",2,3), 
     pch=16,main='Singular Value Decomposition per study',
     xlab=paste("dim1 = ",100*round(res.svd$d^2/sum(res.svd$d^2),3)[1],"%",sep = ""),
     ylab=paste("dim2 = ",100*round(res.svd$d^2/sum(res.svd$d^2),3)[2],"%",sep = ""))
legend("bottomright",legend=c("luad","gbm"), pch=16, cex=0.8, col=2:3)
```

### Quizz

3) Imagine the red and green dots are very far away from each others and each colors cluster togethers. What would be the name of this problem? 

your answer here: 

# Your first classifier

Can you identify the organ of origin from the gene expressions?

*Look at the slides!*

## First step, slit your cohorts.

```{r}
# split
TrainSplit=0.8
Train<-sample(seq(nrow(TCGAdata$CLINIC)),size = round(nrow(TCGAdata$CLINIC)*TrainSplit) )
Val<-seq(nrow(TCGAdata$CLINIC))[!seq(nrow(TCGAdata$CLINIC))%in%Train]
paste("Your training and validation cohorts have",length(Train),"&", length(Val), "patients")
```

### Quizz

4) Why did we split the cohort?

your answer here: 

## Second step, train your logistic regression model.
[glmnet](https://cran.r-project.org/web/packages/glmnet/glmnet.pdf) is a great tool: glm stands for Generalized Linear Models and net for some regularization procedures (optional, set to 0 here). 
The family argument stands for the type of model to use: here you have 2 classes (binomial) and you want to use a traditional logistic regression.

```{r}
# logistic regression fit
LRfit<-glmnet(
  x=as.matrix(
    TCGAdata$EXP[Train,]
  ),
  y=as.factor(
    TCGAdata$CLINIC$study[Train]
  ),
  family="binomial",
  alpha=0,
  lambda=0
)
```

### Quizz

5) On which cohort is it best to evaluate the prediction of your model?

your answer here: 

## Third step, evaluate the predictions on the test set.

```{r}
# preds
Preds<-predict(LRfit,as.matrix(TCGAdata$EXP[Val,]),type = "class")
table(Preds,TCGAdata$CLINIC$study[Val])
```

### Quizz

6) What is the name of such matrix representing true values and predictied values?

your answer here: 

7) What is the [accuracy](https://en.wikipedia.org/wiki/Confusion_matrix) of your current predictions (help in the hyperlink)?

your answer here: 

# Interpret things biologically

Ok, now you have a very good model to classify between lung cancers and glioblastoma from several gene expression, congratulation! 

But maybe you are curious and you want to now what are the genes that are the most important to make such good predictions.

Look at the slides to check out how you can obtain that.

Choose the number of best genes you want to look at (in the cell).

```{r}
Ngenes=20
paste("Your top", Ngenes, "genes are:")
as.matrix(coefficients(LRfit))[head(order(abs(coefficients(LRfit)),decreasing = T),Ngenes),,drop=F]
```

### Quizz

8) Can you say these best 20 gene expressions can determine the tumor type as good as 251?

your answer here: 

# Are there better ways to do so?

Say you want to have a limited list of important genes in your model. It can be useful if a biotech wants to devellop a gene signature to determine the tissue of origin of a cancer sample, for example. It is more or less a regularization (or penalization) procedure: you penalize your model by keeping a limited number of genes.

An existing procedure is Lasso (or L1N) regularization, that increases your cost during training.
*Look at the slides for explanations*

Lasso (like other regularization procedures) comes with an additional parameter to optimize, here $\lambda$. The recommended way to find $\lambda$ is with k-fold cross validation strategy.

```{r}
# choose your lambda with cross validation  (CV)
LRcv<-cv.glmnet(
  x=as.matrix(
    TCGAdata$EXP[Train,]
  ),
  y=as.factor(
    TCGAdata$CLINIC$study[Train]
  ),
  family="binomial",
  alpha=1,
  nfolds=10
)
plot(LRcv)
print(paste("You best value of lambda is :", LRcv$lambda.min))
```

### Quizz

9) Do you think this model needs important regularization to generalize well, and why (max 2 reasons)?

your answer here: 

Then build your final model.

```{r}
# logistic regression fit
LRfitReduced<-glmnet(
  as.matrix(
    TCGAdata$EXP[Train,
    ]
  ),
  as.factor(
    TCGAdata$CLINIC$study[Train]
  ),
  family="binomial",
  alpha=1,
  lambda=LRcv$lambda.min
)
plot(LRfitReduced)
Preds<-predict(LRfitReduced,as.matrix(TCGAdata$EXP[Val,]),type = "class")
table(Preds,TCGAdata$CLINIC$study[Val])
```

### Quizz

10) Has Lasso regularization changed the accuracy of your model?

your answer here: 

Check the selected genes.

```{r}
COEF<-as.matrix(coefficients(LRfitReduced))
GoodCOEF<-COEF[COEF[,1]!=0,,drop=F]
paste("The selected", nrow(GoodCOEF), "genes are:")
GoodCOEF[order(abs(GoodCOEF[,1]),decreasing = T),,drop=F]
```

### Quizz

11) If you go to see a clinician and tell him about your new model, do you think he will want to use it?

your answer here: 

# Produce your own code

Now you have decided to do a clinically usefull model. Your favorite clinician have told you that it would be very usefull to differenciate pancreatic adenocarcinomas and cholangiocarcinomas.

There are new difficulties here: 

- these are rare tumors : probably few examples available publically
- biliary tract and pancreas tumors can be very close to each other from an anotomo-pathological point of view

Tips: [here](https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/tcga-study-abbreviations) you can find the TCGA nomenclatures of tumor types

```{r}
# load TCGA data of pancreatic adenocarcinomas and cholangiocarcinomas 
#### YOUR CODE HERE ####
# set the argument Organ 
TCGAdata<-LoadcBioportal(
  Genes=GENES,
  Organ=c("paad|chol"),
  ClinicNeeded=T,
  MutNeeded=F,
  RNANeeded=T,
  NormalizeRNA=T,
  FunctionalAnnot=F,
  PDF=F,
  Tests=T
)
### END CODE ###
```

How many patients with RNAseq data are there in each studies?

```{r}
#### YOUR CODE HERE #### ~1 line of code
table(TCGAdata$CLINIC$study)
### END CODE ###
```

The distribution of the RNAseq data seems ok to go further with training a relevant model.

```{r}
# Plot distribution
hist(as.matrix(TCGAdata$EXP[TCGAdata$CLINIC$study=="paad",]),col=2,
     main="Distribution of gene expressions",xlab="Normalized TPM", freq = F)
hist(as.matrix(TCGAdata$EXP[TCGAdata$CLINIC$study=="chol",]),add=T,col=3, freq = F)
# Plot SVD in 2D
res.svd<-svd(TCGAdata$EXP, nu = 2, nv = 2)
plot(
    res.svd$v[,1],
    res.svd$v[,2],
    col=ifelse(
	TCGAdata$CLINIC$study=="paad",
	1,
	2
    ), 
    pch=16,
    main='SVD per study',
    xlab=paste("dim1 = ", 100*round(res.svd$d^2/sum(res.svd$d^2),3)[1], "%", sep=""),
    ylab=paste("dim1 = ", 100*round(res.svd$d^2/sum(res.svd$d^2),3)[2], "%", sep=""),
)
legend("bottomright",legend=c("paad","chol"), pch=16, cex=0.8, col=1:2)
```

Now split your cohort into training and test sets.
You will allocate 70% of the cohort in the training.
```{r}
# split
#### YOUR CODE HERE #### ~1 line of code
# give a value to TrainSplit in order to define the split ratio
TrainSplit
### END CODE ###
Train<-sample(seq(nrow(TCGAdata$CLINIC)),size = round(nrow(TCGAdata$CLINIC)*TrainSplit) )
Val<-seq(nrow(TCGAdata$CLINIC))[!seq(nrow(TCGAdata$CLINIC))%in%Train]
length(Train);length(Val)
```

Run cross valiation and retreive the best $\lambda$ value.
```{r}
LRcv<-cv.glmnet(as.matrix(TCGAdata$EXP[Train,]),as.factor(TCGAdata$CLINIC$study[Train]),
                family = "binomial",alpha = 1)
#### YOUR CODE HERE #### ~1 line of code
# print the best lambda value
### END CODE ###
```

Train your final model with the best $\lambda$ value.
```{r}
#### YOUR CODE HERE #### ~1 line of code
# change the lambda argument to train your final model
LRfitReduced<-glmnet(as.matrix(TCGAdata$EXP[Train,]),as.factor(TCGAdata$CLINIC$study[Train]),
              family = "binomial", alpha = 1, lambda = 0)
### END CODE ###
```

Print the confusion matrix on the test set.
```{r}
#### YOUR CODE HERE #### ~2 lines of code
### END CODE ###
```

Retreive the genes used for your model.
```{r}
COEF<-as.matrix(coefficients(LRfitReduced))
GoodCOEF<-COEF[COEF[,1]!=0,,drop=F]
paste("The selected", nrow(GoodCOEF), "genes are:")
GoodCOEF[order(abs(GoodCOEF[,1]),decreasing = T),,drop=F]
```

Congratulations, you have built a classifier that works reasonably well on a clinically difficult task. Nevertheless there are still some work to do before submitting this model to EMA or FDA for approval and commercialisation... To suggest some perspectives, answer to the following questions:

## Questions:

12) Do you think your model needs improvment or it is the best shot ever? 

13) What could be the reasons of some errors of your classifier?

14) What strategies would you suggest to further improve your model?


## Edit your Notebook
You can now edit your notebook to see you results in a nice interface.

---
# A glimpse into survival analysis

The function LoadcBioportal() also loads the survival data from TCGA patients when available. 
Look at the slides to learn what is a survival data.

```{r}
# load luad and gbm data
TCGAdata<-LoadcBioportal(Genes = GENES, Organ = "luad", 
                         ClinicNeeded = T, MutNeeded = F, 
                         RNANeeded = T, NormalizeRNA = T, 
                         FunctionalAnnot = F, PDF = F, Tests=T)
# change status format to numerical
TCGAdata$CLINIC$OS_STATUS<-ifelse(TCGAdata$CLINIC$OS_STATUS=="LIVING",0,1)
```

### Quizz

15) What is the ```TCGAdata$CLINIC$OS_STATUS``` refering to?

your answer here: 

Unfortunatly, there are some NAs and zeros follow up for some patients in TCGA data: we need to discard them.

```{r}
# remove NAS & Zeros follow up
NoNAS<-!is.na(TCGAdata$CLINIC$OS_MONTHS)
table(NoNAS)
NoZeros<-TCGAdata$CLINIC$OS_MONTHS!=0
table(NoZeros)
if(!all(NoNAS&NoZeros)){
  TCGAdata<-lapply(TCGAdata[1:3],function(VAR){
    VAR[NoNAS&NoZeros,]
  })
}
lapply(TCGAdata,dim)
```

Get familiar with the R representation of survival data.
*look at the slides*

```{r}
# look at survival data 
print(Surv(TCGAdata$CLINIC$OS_MONTHS, TCGAdata$CLINIC$OS_STATUS))
```

# Non parametric probability of survival

The ```survfit()``` function of the survival package compute the cumulative probability of survival that takes into acount censored data in the calculation.

At each time step $l$, the cumulative probability $P_{l}$ is calculated by:

$P_{l} = P_{l-1} \cdot \left( \frac{ NatRisk_{l}- Ndeath_{l} }{ NatRisk_{l}} \right)$


```{r}
survival <- survfit(Surv(TCGAdata$CLINIC$OS_MONTHS, TCGAdata$CLINIC$OS_STATUS)~NULL)
summary(survival)
```

The cumulative probability of survival allow you to plot the Kaplan-Meier curve, which is a classical way to visualize the distribution of the survival data accross the cohort.

```{r}
# plot survival data
plot(survfit(Surv(time = TCGAdata$CLINIC$OS_MONTHS, event = TCGAdata$CLINIC$OS_STATUS)~NULL),mark.time = T, 
     main="Kaplan Meier curves per organs", 
     ylab="Probability of survival", xlab="Time in months")
```

# A semi-parametric model survival: the Cox model

Previously your task was to classify the histologies from expression data. Your new task is to predict survival of patients, still from the gene expression.

Modeling survival data for various parameters is a historical discipline (the Cox model has been published in 1952). With the emergence of molecular analysis of cancers it has become a big challenge worldwide. As you will see in this exemple it is much more complicated than histology classification.

Modeling survival looks like a regression task because you have numeric data to predict.
However the way to perform your regression is adapted to the presence of censored data.
The Cox model is using each variable (here gene expression) to model survival data (look at the slides).
*look at the slides*

```{r}
## COX ###
# split
TrainSplit=0.8
Train<-sample(seq(nrow(TCGAdata$CLINIC)),size = round(nrow(TCGAdata$CLINIC)*TrainSplit) )
Val<-seq(nrow(TCGAdata$CLINIC))[!seq(nrow(TCGAdata$CLINIC))%in%Train]
Coxfit<-glmnet(as.matrix(TCGAdata$EXP[Train,]),
               Surv(TCGAdata$CLINIC$OS_MONTHS[Train],TCGAdata$CLINIC$OS_STATUS[Train]),
               family = "cox", alpha = 1, lambda = 0.15)
Preds<-predict(Coxfit,as.matrix(TCGAdata$EXP[Train,]),type = "response")
```


Evaluate the efficacy of your predictions with survival data.
Concordance index is a classical metric (look at the slides). 1 or 0 are perfect concordances and 0.5 is random.
*look at the slides*

```{r}
# check how well your learning has performed on the training set
Conc<-survConcordance(Surv(TCGAdata$CLINIC$OS_MONTHS[Train],TCGAdata$CLINIC$OS_STATUS[Train])~Preds)
plot(TCGAdata$CLINIC$OS_MONTHS[Train],(Preds),pch=TCGAdata$CLINIC$OS_STATUS[Train]+2, col=ifelse(TCGAdata$CLINIC$study[Train]=="luad",1,2))
print(Conc$concordance)
```

```{r}
# check how well your learning has performed on the test set
Preds<-predict(Coxfit,as.matrix(TCGAdata$EXP[Val,]),type = "response")
Conc<-survConcordance(Surv(TCGAdata$CLINIC$OS_MONTHS[Val],TCGAdata$CLINIC$OS_STATUS[Val])~Preds)
plot(TCGAdata$CLINIC$OS_MONTHS[Val],(Preds),pch=TCGAdata$CLINIC$OS_STATUS[Val]+2, col=ifelse(TCGAdata$CLINIC$study[Val]=="luad",1,2))
print(Conc$concordance)
```

The concordance in the training and test sets are both relatively low.

## Quizz bonus

16) What could be the reasons of poor concordance even on the training set?

your answer here:

# End

In this workshop you have learned 2 classical machine learning approaches: classification and regression. We have addressed the challenge of modeling high dimension data (gene expression) to predict clinically relevant observations.

I hope you have gained experience on the good practices in machine learning and the main residing challenges. Maybe these scripts can be usefull for your current or futur research, don't hesitate to use it.

If you want to push foward the adventure, please contact me: loic.verlingue@gustaveroussy.fr . You can also visit our github github.com/DITEP to see what we are working on in our Data Science team.

Thank you!